builtin/packer/config_sourcer.go

```diff
@@ -0,0 +1,235 @@
+)
+
+type ConfigSourcer struct {
+	config sourceConfig
+	client packer.Client
+}
+
+type sourceConfig struct {
+	ClientId       string `hcl:"client_id,optional"`
+	ClientSecret   string `hcl:"client_secret,optional"`
+	OrganizationId string `hcl:"organization_id,attr"`
+	ProjectId      string `hcl:"project_id,attr"`
+}
+
+type reqConfig struct {
+	Bucket  string `hcl:"bucket,attr"`
+	Channel string `hcl:"channel,attr"`
+	Region  string `hcl:"region,attr"`
+	Cloud   string `hcl:"cloud,attr"`
+}
+
+// Config implements component.Configurable
+func (cs *ConfigSourcer) Config() (interface{}, error) {
+	return &cs.config, nil
+}
+
+// ReadFunc implements component.ConfigSourcer
+func (cs *ConfigSourcer) ReadFunc() interface{} {
+	return cs.read
+}
+
+// StopFunc implements component.ConfigSourcer
+func (cs *ConfigSourcer) StopFunc() interface{} {
+	return cs.stop
+}
+
+func (cs *ConfigSourcer) read(
+	ctx context.Context,
+	log hclog.Logger,
+	reqs []*component.ConfigRequest,
+) ([]*pb.ConfigSource_Value, error) {
+
+	// If the user has explicitly set the client ID and secret for the config
+	// sourcer, we use that. Otherwise, we use environment variables.
+	opts := hcpconfig.FromEnv()
+	if cs.config.ClientId != "" && cs.config.ClientSecret != "" {
+		opts = hcpconfig.WithClientCredentials(cs.config.ClientId, cs.config.ClientSecret)
+	}
+	hcpConfig, err := hcpconfig.NewHCPConfig(opts)
+	if err != nil {
+		return nil, err
+	}
+
+	hcpClient, err := httpclient.New(httpclient.Config{
+		HCPConfig: hcpConfig,
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	hcpPackerClient := packer.New(hcpClient, nil)
+	channelParams := packer.NewPackerServiceGetChannelParams()
+	channelParams.LocationOrganizationID = cs.config.OrganizationId
+	channelParams.LocationProjectID = cs.config.ProjectId
+
+	var results []*pb.ConfigSource_Value
+	for _, req := range reqs {
+		result := &pb.ConfigSource_Value{Name: req.Name}
+		results = append(results, result)
+
+		var packerConfig reqConfig
+		// We serialize the config sourcer settings to the reqConfig struct.
+		if err = mapstructure.WeakDecode(req.Config, &packerConfig); err != nil {
+			result.Result = &pb.ConfigSource_Value_Error{
+				Error: status.New(codes.Aborted, err.Error()).Proto(),
+			}
+			return nil, err
+		}
+		channelParams.BucketSlug = packerConfig.Bucket
+		channelParams.Slug = packerConfig.Channel
+
+		// An HCP Packer channel points to a single iteration of a bucket.
+		channel, err := hcpPackerClient.PackerServiceGetChannel(channelParams, nil)
+		if err != nil {
+			return nil, err
+		}
+		log.Debug("Retrieved HCP Packer channel.")
+		iteration := channel.Payload.Channel.Iteration
+
+		// An iteration can have multiple builds, so we check for the first build
+		// with the matching cloud provider and region.
+		for _, build := range iteration.Builds {
+			if build.CloudProvider == packerConfig.Cloud {
+				log.Debug("Found build with matching cloud provider.")
```

Comment:
Same comment here - always nice to include extra relevant info on what was found in these log statements!